rp2/boards: Tidyup Pimoroni boards.

Gadgetoid opened the pull request:
Update- 

I have refactored this PR around the new board variant feature to:

* Remove Tufty 2040, I'll deal with that separately
* Add `MICROPY_DEF_BOARD` to `CMakeLists.txt` as per esp32 (so we can override board names per variant)
* Add 16MB and 4MB variants to Pico LiPo
* Add 2MB and 8MB variants to Tiny 2040

----

For posterity this PR originally said:
* Combine PIMORONI_PICOLIPO_4MB and PIMORONI_PICOLIPO_16MB into a single board using the new variant support for RP2.
* Add 2MB variant for Pimoroni Tiny2040
* Add Tufty2040 and downstream board definition (no support in Pico SDK yet)

Gadgetoid commented:
I'll pop a Tiny 2040 shakeup onto this PR too, hold off merging!

Gadgetoid commented:
Previously we were including the flash size in `MICROPY_HW_BOARD_NAME` which - afaik - is displayed in the REPL.

Switching to variants means we lose that disambiguation. Does anyone have a suggestion for how to add it back? It looks like variants don't load alternate .h files unless they are missing in upstream Pico SDK, in which case the inclusion is just a fallback for "board_name_and_variant.h" missing from from `pico-sdk/tree/master/src/boards/include/boards`.

This means we'd need to use the .cmake file to set some variable (`MICROPY_HW_VARIANT_NAME` ??) that's understood by `CMakeLists.txt`, passed into the build, and merged into the full board name when it's output.

Also this pattern leaves how variant switching is implemented at the behest of the individual boards and feels a little... intangible? It leaves a lot up to the individual board and introduces room for mistakes and hours lost head-scratching. For starters it would be nice if invalid variants were rejected in the same way that invalid boards are.

dpgeorge commented:
@Gadgetoid sorry this got lost, I guess it's because we don't have good answers to your above (very valid) points.

> Previously we were including the flash size in MICROPY_HW_BOARD_NAME which - afaik - is displayed in the REPL.
>
> Switching to variants means we lose that disambiguation. Does anyone have a suggestion for how to add it back?

This is handled on the esp32 port by defining `MICROPY_HW_BOARD_NAME` in a cmake fragment... it's not that elegant but works.  See `ports/esp32/boards/ESP32_GENERIC/mpconfigboard.cmake`, eg:
```cmake
if(MICROPY_BOARD_VARIANT STREQUAL "SPIRAM")
    set(SDKCONFIG_DEFAULTS
        ${SDKCONFIG_DEFAULTS}
        boards/sdkconfig.spiram
    )

    list(APPEND MICROPY_DEF_BOARD
        MICROPY_HW_BOARD_NAME="Generic ESP32 module with SPIRAM"
    )
endif()
```
I suggest to use that mechanism for your rp2 boards, for now.  We might one day improve it.



> Also this pattern leaves how variant switching is implemented at the behest of the individual boards and feels a little... intangible?

Agreed it's a bit flaky / error prone.  You could add some checking in your `mpconfigboard.cmake` file, if the `MICROPY_BOARD_VARIANT` is not a known value then error out.

---------

For now, if you want to rebase this PR on latest master and make the `MICROPY_HW_BOARD_NAME` configurable as described above, then this can be merged.

Gadgetoid commented:
Thank you! I have no recollection of raising this :laughing:  But I'll try and get this done when I can.

Since raising this we switched to using out-of-tree board definitions for all builds (https://github.com/pimoroni/pimoroni-pico/tree/main/micropython/board), so I'll have to consider how this might affect/improve those and how we draw a clear distinction between the official build and... our particular mayhem.

Gadgetoid committed:
ports/rp2: Add MICROPY_DEF_BOARD.

Add MICROPY_DEF_BOARD as per esp32 port, allows board variants to override
the board name with:

list(APPEND MICROPY_DEF_BOARD
    MICROPY_HW_BOARD_NAME="New Board Name"
)

Signed-off-by: Phil Howard <phil@gadgetoid.com>

Gadgetoid committed:
ports/rp2: Refactor Pico LiPo to use variants.

Combine the 4MB and 16MB "PIMORONI_PICOLIPO" variants into a single board.

Signed-off-by: Phil Howard <phil@gadgetoid.com>

Gadgetoid committed:
ports/rp2: Add an 8MB variant to Tiny 2040.

Add an 8MB "PIMORONI_TINY2040" variant.

Signed-off-by: Phil Howard <phil@gadgetoid.com>

Gadgetoid commented:
As promised, I have updated these to reflect the new board variant feature.

Note than in `mpconfigvariant.cmake` I have opted to simply include the "default" variant.

And I have added the fudge to `CMakeLists.txt` so that a variant can override the board name like so:

```cmake
# Override the MicroPython board name
list(APPEND MICROPY_DEF_BOARD
    MICROPY_HW_BOARD_NAME="Pimoroni Tiny 2040 2MB"
)
```

dpgeorge commented:
> Note than in `mpconfigvariant.cmake` I have opted to simply include the "default" variant.

From what I can tell, this means that the `PIMORONI_PICOLIPO` board now has 3 variants:
- the default (equivalent to FLASH_4M)
- FLASH_4M
- FLASH_16M

That means our auto-build system and download page will build the 4M variant twice and list in twice.  This is not ideal.

I guess what you really want is a way to force the specification of a variant, and not have a default variant at all.  But we don't currently support that (in the build system / web site).

I suggest to just remove the `FLASH_4M` from `board.json` so there's only the default (which is 4M) and then the 16M is the only variant.  That would anyway make sense in case a user installs the 4M firmware on a 16M board, it will still work (just have less filesystem).

dpgeorge commented on ports/rp2/boards/PIMORONI_TINY2040/board.json:
```diff
@@ -16,5 +16,9 @@
     "url": "https://shop.pimoroni.com/products/tiny-2040",
+    "variants": {
+        "FLASH_4M": "2 MiB Flash",
+        "FLASH_16M": "8 MiB Flash"
```
I think the keys here should be `"FLASH_2M"` and `"FLASH_4M"`.

dpgeorge commented:
Thanks for updating.  Rebased and merged in 11becbe22321ca1d7bd631144976f246175730ce through 1557014ea4e88a1dfbef028afa393e6d2c3e6a9e
